### PR TITLE
Autofill!

### PR DESCRIPTION
**Pull request by sashei**
Connected to #486

With the addition of the Credential Provider API in iOS 12, you will need to use **Xcode 10** (beta 6) and **Swift 4.2** (`swiftlang-1000.0.32.1 clang-1000.10.39`) to work on or build this branch.

This will serve as the long-running feature branch until we can work on `master` directly due to Xcode upgrades.

Remaining todos:

- [x] test and merge #666

### CODE REVIEW

**Comment by devinreams** [comment]
Oh yes! Please tell me you'd like to do Pull Requests _into this branch_ for all the other little feature branches 🤞 I love branching strategy...

**Comment by sashei** [comment]
as much as it pains me to say it.... yes.

**Comment by sashei** [comment]
@irrationalagent the `CredentialProviderPresenter.swift` is the rough equivalent of the `RootPresenter` in Lockbox.app. There is a similar `Action` cycle in this application though, I might hold tight for the next PR to come in (#649) because that will have more infrastructure for Telemetry-interesting actions in place.

File: CredentialProvider/View/CredentialProviderView.swift [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     Prepare your UI to list available credentials for the user to choose from. The items in
+     'serviceIdentifiers' describe the service the user is logging in to, so your extension can
+     prioritize the most relevant credentials in the list.
+    */
+    override func prepareCredentialList(for serviceIdentifiers: [ASCredentialServiceIdentifier]) {
+        
+    }
+
+    override func provideCredentialWithoutUserInteraction(for credentialIdentity: ASPasswordCredentialIdentity) {
+        let databaseIsUnlocked = true
```
Comment (jhugman) [inline CredentialProvider/View/CredentialProviderView.swift]:
```swift
// Dummy implementation
```

File: Shared/Common/Extensions/ASCredentialIdentityStore+.swift [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import AuthenticationServices
+
+@available(iOS 12, *)
+protocol CredentialIdentityStoreProtocol {
+    func getState(_ completion: @escaping (ASCredentialIdentityStoreState) -> Void)
+    func removeAllCredentialIdentities(_ completion: ((Bool, Error?) -> Void)?)
```
Comment (jhugman) [inline Shared/Common/Extensions/ASCredentialIdentityStore+.swift]: Suggest using a typealias here.
e.g. 

```swift
typealias CredentialCompletion = ((Bool, Error?) -> Void)
```

Then use it as optional in the method parameter.

```swift
func removeAllCredentialIdentities(_ completion: CredentialCompletion?)
```

File: Shared/Store/BaseDataStore.swift [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+        case (_, _, NotificationNames.ProfileDidStartSyncing):
+            // fall through for the locked and unlocked states.
+            syncSubject.onNext(.Syncing)
+        case (_, _, NotificationNames.ProfileDidFinishSyncing):
+            // end of all syncs
+            syncSubject.onNext(.Synced)
+        default:
+            print("Unexpected state combination: \(storageState) | \(syncState), \(notification.name)")
+        }
+    }
```
Comment (jhugman) [inline Shared/Store/BaseDataStore.swift]: Cherrypicking is going to be annoying here.

File: Shared/Store/BaseDataStore.swift [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+        }
+
+        func resetProfile() {
+            self.profile = profileFactory(true)
+            self.initializeProfile()
+            self.syncSubject.onNext(.NotSyncable)
+            self.storageStateSubject.onNext(.Unprepared)
+        }
+
+        stopSyncing() >>== disconnect >>== deleteAll >>== resetProfile
```
Comment (jhugman) [inline Shared/Store/BaseDataStore.swift]: Cherrypicking is going to be annoying here.

File: Shared/Store/BaseDataStore.swift [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+                .subscribe(onNext: { autoLockSetting in
+                    switch autoLockSetting {
+                    case .Never:
+                        self.unlock()
+                    default:
+                        let date = NSDate(
+                                timeIntervalSince1970: self.userDefaults.double(
+                                        forKey: UserDefaultKey.autoLockTimerDate.rawValue))
+
+                        if date.timeIntervalSince1970 > 0 && date.timeIntervalSinceNow > 0 {
```
Comment (jhugman) [inline Shared/Store/BaseDataStore.swift]: Prefer to remove `>`.

oh, wait this is to make sure we have an `autoLockTimerDate`.

```swift
if let seconds = self.userDefaults.double(forKey: UserDefaultKey.autoLockTimerDate.rawValue),
     seconds > 0 {

} else {

}
```

Actually, I'm not sure what this code is supposed to do, or that it wasn't looked at in another PR.

File: lockbox-ios/Store/AccountStore.swift [hunk 6]
```diff
@@ -8,57 +8,65 @@ import RxCocoa
...
-    static let allValues: [KeychainKey] = [.accountJSON, .email, .displayName, .avatarURL]
+import Shared
+
+/* These UserDefault keys are maintained separately from the `UserDefaultKey`
+ * enum in BaseConstants.swift because, for the time being, they are only
+ * useful in the main application. However, rather than maintaining two
+ * separate UserDefaults instances, all the values for these keys are stored
+ * in the app group instance so that no further migrations will be required
+ * in the case that they become relevant in an app extension context. */
+enum LocalUserDefaultKey: String {
```
Comment (jhugman) [inline lockbox-ios/Store/AccountStore.swift]: I'm not sure I understand the need for two sets of keys, especially if there is scope for sharing the keys.

**Comment by devinreams** [comment]
Quick status update:

- three uispecs are failing ❌ but known and understood 👌 
  - two of them (3 and 6) are kinda flakey and being addressed at #686 
  - one is because we introduced a screen before the welcome (0)

- @jhugman has requested some nits/changes

- shortly after this message, this branch will require **beta 6 of Xcode 10** (up from beta 5)

File: CredentialProvider/Presenter/CredentialWelcomePresenter.swift [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    private func populateCredentials() {
+        self.dispatcher.dispatch(action: CredentialProviderAction.refresh)
+
+        let populated = self.credentialProviderStore.state
+            .filter { $0 == CredentialProviderStoreState.Populated }
+            .map { _ -> Void in return () }
+
+        self.credentialProviderStore.state
+                .asDriver(onErrorJustReturn: CredentialProviderStoreState.NotAllowed)
```
[the remark this answers is not in the document]
Comment (sashei) [reply CredentialProvider/Presenter/CredentialWelcomePresenter.swift]: It's used later for onboarding ... I almost did though, you're right.

File: Shared/Common/Extensions/ASCredentialIdentityStore+.swift (hunk 2 above)
Comment (sashei) [reply Shared/Common/Extensions/ASCredentialIdentityStore+.swift]: doing this messes up the injection pattern that these are useful for :(

File: Shared/Store/BaseDataStore.swift (hunk 5 above)
Comment (sashei) [reply Shared/Store/BaseDataStore.swift]: this is from another PR.

File: lockbox-ios/Store/AccountStore.swift (hunk 6 above)
Comment (sashei) [reply lockbox-ios/Store/AccountStore.swift]: It has to do with minimizing what settings are included in the extension. eg, we don't need to share the preferred browser setting in the extension.

**Pull request merged**